### PR TITLE
feature/add support for public eks clusters

### PR DESCRIPTION
**Pull request by omissis**
- chore(bootstrap): introduce support for independent vpc and vpn modules
- chore(cluster): introduce support for public and public+private clusters

### CODE REVIEW

File: data/provisioners/bootstrap/aws/main.tf [hunk 1]
```diff
@@ -9,26 +9,43 @@ terraform {
...
+  tags         = var.tags
+
+  private_subnetwork_cidrs = var.vpc_private_subnetwork_cidrs
+  public_subnetwork_cidrs  = var.vpc_public_subnetwork_cidrs
+}
+
+module "vpn" {
+  source = "github.com/sighupio/fury-eks-installer//modules/vpn?ref=v2.0.0"
+
+  count = var.vpn_enabled ? 1 : 0
```
Comment (ralgozino) [inline data/provisioners/bootstrap/aws/main.tf:33]: with this approach, we are losing the capability of having more than one bastion. In production environments, we usually deploy 2 instances of the Bastion (VPN) to have HA.

Instead of hardcoding `1` we should use the value of `vpn_instances`

File: data/provisioners/bootstrap/aws/variables.tf [hunk 2]
```diff
@@ -78,6 +105,7 @@ variable "vpn_operator_cidrs" {
 }
 
 variable "vpn_ssh_users" {
-  description = "GitHub users id to sync public rsa keys. Example angelbarrera92"
+  description = "GitHub users to sync public rsa keys"
```
Comment (ralgozino) [inline data/provisioners/bootstrap/aws/variables.tf:108]:
```suggestion
  description = "GitHub users to sync public keys for SSH access"
```

**Review by ralgozino — CHANGES_REQUESTED** [review]

File: internal/bootstrap/provisioners/aws/provisioner.go [hunk 3]
```diff
@@ -25,12 +25,11 @@ func (d *AWS) InitMessage() string {
...
 
 This provisioner creates a battle-tested AWS VPC with all the requirements
-set to run a production-grade private EKS cluster.
+set to run a production-grade EKS cluster.
 
-It creates a VPN server enables deploying the cluster from this computer
-once connected to the VPN server.
-
-Then, use furyagent to manage VPN profiles.
+If you opt to create a private cluster, the provisioner will create a VPN Server that will
```
[the remark this answers is not in the document]
Comment (omissis) [reply internal/bootstrap/provisioners/aws/provisioner.go:30]: You're right, it won't. We should probably not force that behaviour though(as some will want their own vpn and a private cluster), so it's better to rewrite this phrase. wdyt?

File: data/provisioners/bootstrap/aws/main.tf (hunk 1 above)
Comment (omissis) [reply data/provisioners/bootstrap/aws/main.tf:33]: hmm not really, the module hasn't fundamentally changed from the previous version, besides being split. we are still able to create multiple instances via the `vpn_instances` variable down below, pretty much as (I assume) we did in the past.
The semantic here is simply on/off. If it's on, then you can configure it exactly as we used to do in the past (that is, using multiple instances if deemed necessary).

File: data/provisioners/bootstrap/aws/main.tf [hunk 4]
```diff
@@ -9,26 +9,43 @@ terraform {
...
+
+  name         = var.name
+  network_cidr = var.network_cidr
+  tags         = var.tags
+
+  vpc_id         = var.vpc_enabled ? one(module.vpc[*].vpc_id) : var.vpn_vpc_id
+  public_subnets = var.vpc_enabled ? one(module.vpc[*].public_subnets) : var.vpn_public_subnets
```
[the remark this answers is not in the document]
Comment (omissis) [reply data/provisioners/bootstrap/aws/main.tf:40]: it's not, it's passing the list of subnets as it did before, and then the resource will point to the correct one by index.
The difference lays in the fact that earlier we only received this list from the `module.vpc`, whereas now we can also pass it manually, should the user desire to use their own vpc.

File: data/provisioners/bootstrap/aws/main.tf (hunk 1 above)
Comment (ralgozino) [reply data/provisioners/bootstrap/aws/main.tf:33]: ohh I think I understand, the `count` affects only the count of `vpn` modules being created and not the number of instances, right?

File: data/provisioners/bootstrap/aws/main.tf (hunk 4 above)
[the remark this answers is not in the document]
Comment (ralgozino) [reply data/provisioners/bootstrap/aws/main.tf:40]: the usage of `one()` confuses me here, but I believe you :)

File: internal/bootstrap/provisioners/aws/provisioner.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ralgozino) [reply internal/bootstrap/provisioners/aws/provisioner.go:30]: Yeap I agree, VPN creation should be opt-in. I'd rephrase this to something like: "[...] the provisioner CAN create a VPN server for you that will[...], you can enable it by setting <blah blah>"

File: data/provisioners/bootstrap/aws/main.tf (hunk 1 above)
Comment (omissis) [reply data/provisioners/bootstrap/aws/main.tf:33]: correct 👍

File: data/provisioners/bootstrap/aws/main.tf (hunk 4 above)
[the remark this answers is not in the document]
Comment (omissis) [reply data/provisioners/bootstrap/aws/main.tf:40]: yeah, the usage of `one()` is required by the use of `count` in the vpc module: it being present makes terraform believe the vpc module could return a list of objects, and therefore I can't write `module.vpc.vpc_id` anymore due to type mismatch.
Practically speaking though, module.vpc will always be a 0 or 1 item list, but in order to make terraform correctly access the values, we need to treat module.vpc as a list and get the first item

[the remark this answers is not in the document]
Comment (ralgozino) [reply data/provisioners/bootstrap/aws/main.tf:40]: gotcha, thanks!

**Comment by omissis** [comment]
converting to draft as we are not sure we are going to release this feature on legacy furyctl

**Comment by omissis** [comment]
Closing as we'll likely won't support public clusters in furyctl legacy
